Remove default logging scope from ASP.NET Core

davidfowl opened the pull request:
- We log the activity information by default, the span id is more useful than the traceidentifier.
- Add the path and trace id as activity tags

I was looking at this and wondering why we still need it other than compatibility. I'm thinking we can lean into activity (as big as it is 😄) and use it's scope instead of having 2 by default. This reduces the number of async locals as well which is a plus. @noahfalk do you think it's reasonable to have a feature for logging adding tags to the logging activity scope as well?

cc @noahfalk @tarekgh @adityamandaleeka 

These are the per request allocations (excluding execution context overhead) (10K requests):

|Type|Allocations|Bytes|Average Size \(Bytes\)|
|-|-|-|-|
|\| - Microsoft.Extensions.Logging.Logger.Scope|10,125|486,000|48|
|\| - Microsoft.Extensions.Logging.LoggerFactoryScopeProvider.Scope|10,125|486,000|48|
|\| - Microsoft.AspNetCore.Hosting.HostingApplicationDiagnostics.Log.HostingLogScope|10,000|400,000|40|
|\| - System.IDisposable\[\]|10,125|324,000|32|

tarekgh commented:
@davidfowl I think this is change is good if we are not concering about the compat. If we proceed with such change would be good to file breaking change doc for that.

CC @reyang @cijothomas

cijothomas commented on src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:
```diff
@@ -65,6 +63,11 @@ public void BeginRequest(HttpContext httpContext, HostingApplication.Context con
             if (context.Activity is Activity activity)
             {
+                // Add some ASP.NET Core specific properties as tags
+                activity.AddTag("aspnet.traceidentifier", httpContext.TraceIdentifier);
```
most of the time, people do sampling with activity, so if the activity is sampled, then they'd lose this information... do you see that as a concern?

davidfowl replied:
I'm not sure. This also affects activity log correlation as well right?

davidfowl replied:
I've removed these completely.

noahfalk commented on src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:
```diff
@@ -88,11 +91,6 @@ public void BeginRequest(HttpContext httpContext, HostingApplication.Context con
-            // Scope may be relevant for a different level of logging, so we always create it
-            // see: https://github.com/aspnet/Hosting/pull/944
-            // Scope can be null if logging is not on.
-            context.Scope = Log.RequestScope(_logger, httpContext);
```
I think my main concern about switching from logger scope to Activity is checking whether typical tooling that users have available today for log viewing can do a good job correlating Activity tags with log messages. There is no reason that tools couldn't do it, but it may be something that wasn't important previously so the tools didn't invest the energy into it. On the one hand if do this then it will provide motivations for tools to support it better, but it might leave users in a worse place until the tools catch up. A more conservative route would be adding the new info to Activity and tell tooling folks that we plan to remove the data from logger scope in the future to encourage them to prepare.

davidfowl replied:
I think 2 things can help mitigate the breaking change:
1. We support pushing activity tags to the logging activity scope (the same way we do for the built in ones)
2. Assuming we did this, it would flow to logger provider implementations that supported the IExternalScopeProvider

But it would mean that we aren't explicitly adding a logging scope so if the ILoggerProvider doesn't support IExternalScopeProvider, it would be more visibly breaking.

noahfalk replied:
I don't have any background knowledge on how often users correlate log messages using those tags we are moving, so I'm mostly relying on you to decide how breaking this is. I could see how an option that includes all Activity tags in ExternalScopeProvider could be useful, but if you think making this change safe requires support for that option first then we probably want to put this PR on hold until then.

davidfowl replied:
Agree, I'll leave it as a draft. I think we should open up a new issue for "Activity tags as log scope entries".

tarekgh replied:
We already support logging Activity tags in https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.activitytrackingoptions?view=dotnet-plat-ext-6.0

davidfowl replied:
Ah, nice!

davidfowl replied:
I'm pretty convinced this scope useful by default and can be easily added back via middleware. The cost we're paying for the benefit people get by default is too high (many don't even know how to look at these logging scopes). I think now that the ability to associate logs with the activity exists and is on by default. I feel pretty confident breaking this.

noahfalk commented:
> @noahfalk do you think it's reasonable to have a feature for logging adding tags to the logging activity scope as well?

what if we encourage people to phase out usage of logger scope and use Activity instead? My worry if we try to automatically migrate tags is that the scope lifetimes may not match up and you could have weird results. For example:

```C#
// this might be ASP.Net code
void HandleRequest()
{
    using (Activity a = source.StartActivity)
    {
        UserRequestHandler(requestData);
    }
}

void UserRequestHandler(requestData)
{
    foreach(var item in requestData.Batch)
    {
        // this scope doesn't have 1:1 relationship with Activity, the tags would overwrite each other
        using(logger.CreateScope("Id", item.Id))
        {
            logger.Log("stuff");
        }
        // this log is outside the logger scope, but still inside the Activity scope where
        // the tags have now been forwarded.
        logger.Log("all done now");
    }
}
```

NinoFloris commented:
Relevant https://github.com/dotnet/aspnetcore/issues/31028

davidfowl commented on src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:
```diff
@@ -65,6 +63,11 @@ public void BeginRequest(HttpContext httpContext, HostingApplication.Context con
             if (context.Activity is Activity activity)
             {
+                // Add some ASP.NET Core specific properties as tags
+                activity.AddTag("aspnet.traceidentifier", httpContext.TraceIdentifier);
```
This is really only useful in the IIS and HTTP.sys. They flow an ID that is can be used to correlate with their lower level ETW traces.

noahfalk replied:
Do you think we could pick a name that is more IIS/HTTP.sys specific and only include the identifier when asp.net is run in that hosting environment?

davidfowl replied:
Is the ASP.NET prefix not sufficient for that? Kestrel invents its own ID today as well. There's no easy way from here to tell whether we're running in IIS or HTTP.sys. Since the activity is created later then the server can detect, it can't do anything by default since it doesn't have that reference. 

Another possibility is we don't add anything, and developers can write code to enrich the activity with the trace identifier when they need this functionality. Not ideal but the only places we get access to the http context and the activity is middleware or our `OnActivityImport` diagnostic source event.

davidfowl replied:
I've removed it.

davidfowl commented on src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs:
```diff
@@ -65,6 +63,11 @@ public void BeginRequest(HttpContext httpContext, HostingApplication.Context con
+                // Add some ASP.NET Core specific properties as tags
+                activity.AddTag("aspnet.traceidentifier", httpContext.TraceIdentifier);
+                var path = (httpContext.Request.PathBase.HasValue ? httpContext.Request.PathBase + httpContext.Request.Path : httpContext.Request.Path).ToString();
+                activity.AddTag("aspnet.requestpath", path);
```
Sounds good to me.

davidfowl commented:
/azp run

amcasey commented:
@davidfowl What's happening with this?

davidfowl committed:
Stop creating the default logging scope in hosting
- We correlate logs with Activities and more information can be added to them (like the trace identifier) using activity listeners or custom middleware.

davidfowl commented:
@noahfalk I updated this PR to just remove the scopes. 

I'm pretty convinced this scope useful by default and can be easily added back via middleware. The cost we're paying for the benefit people get by default is too high (many don't even know how to look at these logging scopes). I think now that the ability to associate logs with the activity exists and is on by default. I feel pretty confident breaking this.

We will show people how to add a middleware back and adds this information via a logging scope or activity tags.

joegoldman2 commented:
Can we expect this change for .NET 9? What about Microsoft.AspNetCore.Server.Kestrel.Core.Internal.ConnectionLogScope?
